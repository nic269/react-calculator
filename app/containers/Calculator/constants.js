/*
 *
 * Calculator constants
 *
 */
import _flattenDeep from 'lodash/flattenDeep';

export const GET_RESULT = 'app/Calculator/GET_RESULT';
export const CLEAR_RESULT = 'app/Calculator/CLEAR_RESULT';
export const DELETE_AG = 'app/Calculator/DELETE_AG';
export const UPDATE_FORMULA = 'app/Calculator/UPDATE_FORMULA';

export const ADD_KEY = '+';
export const SUB_KEY = '-';
export const MUL_KEY = '*';
export const DIV_KEY = '/';
export const CLEAR_GROUP = 'clear';
export const DELETE_GROUP = 'delete';
export const OPERATOR_GROUP = 'operator';
export const SUBMIT_GROUP = 'submit';
export const NUMBER_GROUP = 'number';
export const OPERATION_KEYS = [ADD_KEY, SUB_KEY, MUL_KEY, DIV_KEY];
export const SUBMIT_KEYS = ['Enter', '='];
export const CLEAR_KEYS = ['Escape'];
export const DELETE_KEYS = ['Backspace'];
export const LAYOUT = [
  [
    {
      id: '1',
      name: 'AC',
      group: CLEAR_GROUP,
      size: 'medium',
      key: 'Escape',
      description: 'Clear all (or press Esc)',
    },
    {
      id: '1_1',
      name: '⇤',
      group: DELETE_GROUP,
      key: 'Backspace',
      description: 'Delete (or press Backspace)',
    },
    {
      id: '2',
      name: '/',
      group: OPERATOR_GROUP,
      key: DIV_KEY,
      description: 'Divide (or press /)',
    },
  ],
  [
    {
      id: '3',
      name: '1',
      group: NUMBER_GROUP,
      key: '1',
    },
    {
      id: '4',
      name: '2',
      group: NUMBER_GROUP,
      key: '2',
    },
    {
      id: '5',
      name: '3',
      group: NUMBER_GROUP,
      key: '3',
    },
    {
      id: '6',
      name: '*',
      group: OPERATOR_GROUP,
      key: MUL_KEY,
      description: 'Multiply (or press *)',
    },
  ],
  [
    {
      id: '7',
      name: '4',
      group: NUMBER_GROUP,
      key: '4',
    },
    {
      id: '8',
      name: '5',
      group: NUMBER_GROUP,
      key: '5',
    },
    {
      id: '9',
      name: '6',
      group: NUMBER_GROUP,
      key: '6',
    },
    {
      id: '10',
      name: '-',
      group: OPERATOR_GROUP,
      key: SUB_KEY,
      description: 'Subtract (or press -)',
    },
  ],
  [
    {
      id: '11',
      name: '7',
      group: NUMBER_GROUP,
      key: '7',
    },
    {
      id: '12',
      name: '8',
      group: NUMBER_GROUP,
      key: '8',
    },
    {
      id: '13',
      name: '9',
      group: NUMBER_GROUP,
      key: '9',
    },
    {
      id: '14',
      name: '+',
      group: OPERATOR_GROUP,
      key: ADD_KEY,
      description: 'Add (or press +)',
    },
  ],
  [
    {
      id: '15',
      name: '0',
      group: NUMBER_GROUP,
      size: 'large',
      key: '0',
    },
    {
      id: '16',
      name: '＝',
      group: SUBMIT_GROUP,
      key: '=',
      description: 'Equal (or press Enter)',
    },
  ],
];
export const SUPPORTED_KEYS = [
  ..._flattenDeep(LAYOUT).map(item => item.key),
  'Enter',
  'Backspace',
];
