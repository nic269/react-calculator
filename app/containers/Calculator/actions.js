/*
 *
 * Caculator actions
 *
 */

import { GET_RESULT, CLEAR_RESULT, UPDATE_FORMULA } from './constants';

export const getResult = () => ({
  type: GET_RESULT,
});

export const clearResult = () => ({
  type: CLEAR_RESULT,
});

export const updateFormula = payload => ({
  type: UPDATE_FORMULA,
  payload,
});
