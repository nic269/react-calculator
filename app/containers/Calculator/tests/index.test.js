/**
 *
 * Tests for Caculator
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import configureStore from '../../../configureStore';
import { Calculator, mapDispatchToProps } from '../index';
import { getResult, clearResult, deleteAg, updateFormula } from '../actions';
import { NUMBER_GROUP, OPERATOR_GROUP } from '../constants';

jest.mock('components/KeyPad', () => 'MockedKeyPad');
jest.mock('components/Screen', () => 'MockedScreen');

describe('<Caculator />', () => {
  let store;
  let props;

  beforeAll(() => {
    props = {
      calculatorHandler: jest.fn(),
    };
    store = configureStore({}, browserHistory);
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Calculator {...props} />
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('Should not update calculator when user press keyboard with key in supported list', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Calculator {...props} />
      </Provider>,
    );
    fireEvent.keyDown(firstChild, { key: 'A', code: 65 });

    expect(props.calculatorHandler).not.toHaveBeenCalled();
  });

  it('Should update calculator when user press keyboard with key in supported list', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Calculator {...props} />
      </Provider>,
    );
    fireEvent.keyDown(firstChild, { key: 'Enter', code: 13 });

    expect(props.calculatorHandler).toHaveBeenCalledWith('Enter');
  });
});

describe('mapDispatchToProps', () => {
  describe('calculatorHandler', () => {
    it('should be injected', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      expect(result.calculatorHandler).toBeDefined();
    });

    it('should call getResult', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.calculatorHandler('Enter');
      expect(dispatch).toHaveBeenCalledWith(getResult());
    });

    it('should call clearResult', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.calculatorHandler('Escape');
      expect(dispatch).toHaveBeenCalledWith(clearResult());
    });

    it('should call deleteAg', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.calculatorHandler('Backspace');
      expect(dispatch).toHaveBeenCalledWith(deleteAg());
    });

    it('should call updateFormula with number', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.calculatorHandler('1');
      expect(dispatch).toHaveBeenCalledWith(
        updateFormula({
          value: '1',
          group: NUMBER_GROUP,
        }),
      );
    });

    it('should call updateFormula with operator', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.calculatorHandler('+');
      expect(dispatch).toHaveBeenCalledWith(
        updateFormula({
          value: '+',
          group: OPERATOR_GROUP,
        }),
      );
    });
  });
});
