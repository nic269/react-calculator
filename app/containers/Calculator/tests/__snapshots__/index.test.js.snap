// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Caculator /> Should render and match the snapshot 1`] = `
<div>
  <mockedscreen
    value="0"
  />
  <mockedkeypad
    layout="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
  />
</div>
`;
